src/endpoint/s3/s3_bucket_policy_utils.js

```diff
@@ -189,7 +201,23 @@ function _is_principal_fit(account, statement) {
             break;
         }
     }
-    return statement.Principal ? principal_fit : !principal_fit;
+
+    if (statement.NotPrincipal) {
+        // If the account is categorized as "Not Principal," we need to ALLOW operations
```

Comment:
have you looked at https://docs.aws.amazon.com/IAM/latest/UserGuide/reference_policies_elements_notaction.html?
note that there is an implicit Deny for actions that does not any statement. combination of Deny with NotPrincipal is not the same as Allow